libs/vkd3d/swapchain.c

```diff
@@ -2444,6 +2682,181 @@ static CONST_VTBL struct IDXGIVkSwapChainFactoryVtbl dxgi_vk_swap_chain_factory_
     dxgi_vk_swap_chain_factory_CreateSwapChain,
 };
 
+bool dxgi_vk_swap_chain_low_latency_enabled(struct dxgi_vk_swap_chain* chain)
+{
+    return chain->present.low_latency_state.mode;
+}
+
+void dxgi_vk_swap_chain_latency_sleep(struct dxgi_vk_swap_chain* chain)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &chain->queue->device->vk_procs;
+    VkLatencySleepInfoNV latency_sleep_info;
+    VkSemaphoreWaitInfo sem_wait_info;
+    bool should_sleep = false;
+
+    /* Increment the low latency sem value before the wait */
+    chain->present.low_latency_sem_value++;
+
+    memset(&latency_sleep_info, 0, sizeof(latency_sleep_info));
+    latency_sleep_info.sType = VK_STRUCTURE_TYPE_LATENCY_SLEEP_INFO_NV;
+    latency_sleep_info.pNext = NULL;
+    latency_sleep_info.signalSemaphore = chain->present.low_latency_sem;
+    latency_sleep_info.value = chain->present.low_latency_sem_value;
+
+    pthread_mutex_lock(&chain->present.low_latency_swapchain_lock);
+
+    if (chain->present.vk_swapchain)
+    {
+        should_sleep = true;
+        VK_CALL(vkLatencySleepNV(chain->queue->device->vk_device, chain->present.vk_swapchain, &latency_sleep_info));
+    }
+
+    pthread_mutex_unlock(&chain->present.low_latency_swapchain_lock);
+
+    if (should_sleep)
+    {
+        memset(&sem_wait_info, 0, sizeof(sem_wait_info));
+        sem_wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO;
+        sem_wait_info.pNext = NULL;
+        sem_wait_info.flags = 0;
+        sem_wait_info.semaphoreCount = 1;
+        sem_wait_info.pSemaphores = &chain->present.low_latency_sem;
+        sem_wait_info.pValues = &chain->present.low_latency_sem_value;
+
+        VK_CALL(vkWaitSemaphores(chain->queue->device->vk_device, &sem_wait_info, UINT64_MAX));
+    }
+}
+
+void dxgi_vk_swap_chain_set_latency_sleep_mode(struct dxgi_vk_swap_chain* chain, bool low_latency_mode,
+	bool low_latency_boost, uint32_t minimum_interval_us)
+{
+    pthread_mutex_lock(&chain->present.low_latency_state_update_lock);
```

Comment:
Is the state update lock actually needed now?

Reply:
I thought it would be good to be robust against race conditions between Present, and SetLatencySleepMode. Given the lock shouldn't be significantly contended the overhead should be small.

Reply:
Fair enough.